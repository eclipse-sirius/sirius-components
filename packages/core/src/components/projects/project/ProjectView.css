/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

.projectview {
  display: grid;
  grid-template-rows: min-content auto;
  grid-template-columns: 1fr;
  grid-row-gap: var(--layoutDimension-l);
}

.projectview-main {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: 1fr 1fr;
  grid-column-gap: var(--layoutDimension-l);
}

.projectview-details .card {
  margin-top: var(--layoutDimension-l);
}
.projectview-details .card:nth-child(1) {
  margin-top: 0;
}
