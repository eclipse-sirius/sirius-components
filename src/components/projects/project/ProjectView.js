/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

import React from 'react';
import PropTypes from 'prop-types';

import { classNames } from '../../../common/classnames';

import { Loading } from '../../loading/Loading';

import { ProjectHeaderCard } from '../ProjectHeaderCard';
import { ProjectRepresentationsListCard } from '../ProjectRepresentationsListCard';
import { ProjectSemanticResourcesListCard } from '../ProjectSemanticResourcesListCard';

import { LOADING__STATE, PROJECT_LOADED__STATE } from './ProjectViewFiniteStateMachine';

import './ProjectView.css';

const propTypes = {
  stateId: PropTypes.string
};

/**
 * The ProjectView is used to display and manipulate a project.
 */
export const ProjectView = ({ className, stateId, error, project, ...props }) => {
  switch (stateId) {
    case LOADING__STATE:
      return renderLoadingState(className, props);
    case PROJECT_LOADED__STATE:
      return renderProjectLoadedState(className, project, props);
    default:
      return renderLoadingState(className, props);
  }
};
ProjectView.propTypes = propTypes;

const renderLoadingState = (className, props) => <Loading className={className} {...props} />;

const PROJECT_VIEW__CLASS_NAMES = 'projectview';
const PROJECT_VIEW_MAIN__CLASS_NAMES = 'projectview-main';
const PROJECT_VIEW_DETAILS__CLASS_NAMES = 'projectview-details';

const renderProjectLoadedState = (className, project, props) => {
  const projectViewClassNames = classNames(PROJECT_VIEW__CLASS_NAMES, className);
  return (
    <div className={projectViewClassNames}>
      <ProjectHeaderCard name={project.name} />
      <div className={PROJECT_VIEW_MAIN__CLASS_NAMES}>
        <div className={PROJECT_VIEW_DETAILS__CLASS_NAMES}>
          <ProjectSemanticResourcesListCard semanticResources={project.semanticResources} />
          <ProjectRepresentationsListCard representations={project.representations} />
        </div>
      </div>
    </div>
  );
};
