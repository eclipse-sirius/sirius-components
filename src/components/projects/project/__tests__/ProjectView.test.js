/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import Renderer from 'react-test-renderer';

import { ProjectView } from '../ProjectView';

import {
  ERROR__STATE,
  INITIAL__STATE,
  LOADING__STATE,
  PROJECT_LOADED__STATE
} from '../ProjectViewFiniteStateMachine';

describe('ProjectView', () => {
  it('renders an error card for an error', () => {
    const error = {
      title: 'An error has occurred',
      message: 'Please contact the administrator to try to find a solution',
      code: 500
    };
    const project = Renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <ProjectView stateId={ERROR__STATE} error={error} project={null} />
      </MemoryRouter>
    );
    expect(project.toJSON()).toMatchSnapshot();
  });

  it('renders an error card for an unsupported state', () => {
    const project = Renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <ProjectView stateId={INITIAL__STATE} error={null} project={null} />
      </MemoryRouter>
    );
    expect(project.toJSON()).toMatchSnapshot();
  });

  it('renders a loading indicator during the loading', () => {
    const project = Renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <ProjectView stateId={LOADING__STATE} error={null} project={null} />
      </MemoryRouter>
    );
    expect(project.toJSON()).toMatchSnapshot();
  });

  it('renders the project', () => {
    const project = {
      name: 'Project',
      representations: [{ name: 'Class Diagram' }, { name: 'Activity Diagram' }],
      semanticResources: [{ name: 'model.ecore', path: '/model/model.ecore' }]
    };
    const projectComponent = Renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <ProjectView stateId={PROJECT_LOADED__STATE} error={null} project={project} />
      </MemoryRouter>
    );
    expect(projectComponent.toJSON()).toMatchSnapshot();
  });
});
