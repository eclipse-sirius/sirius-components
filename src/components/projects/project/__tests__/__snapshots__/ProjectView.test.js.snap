// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectView renders a loading indicator during the loading 1`] = `
<div
  className=""
>
  Loading
</div>
`;

exports[`ProjectView renders an error card for an error 1`] = `
<div
  className="card errorcard"
>
  <div
    className="card-body"
  >
    <h1
      className="title-l"
    >
      An error has occurred
    </h1>
    <p>
      Please contact the administrator to try to find a solution (code: 500)
    </p>
  </div>
</div>
`;

exports[`ProjectView renders an error card for an unsupported state 1`] = `
<div
  className="card errorcard"
>
  <div
    className="card-body"
  >
    <h1
      className="title-l"
    >
      The project is in an unsupported state: INITIAL__STATE
    </h1>
    <p>
      Contact your administrator to find a suitable solution (code: 1001)
    </p>
  </div>
</div>
`;

exports[`ProjectView renders the project 1`] = `
<div
  className="projectview"
>
  <div
    className="card projectheadercard"
  >
    <div
      className="card-header"
    >
      <h1
        className="card-title title-xl"
      >
        Project
      </h1>
    </div>
  </div>
  <div
    className="projectview-main"
  >
    <div
      className="projectview-details"
    >
      <div
        className="card projectsemanticresourceslistcard"
      >
        <div
          className="card-header"
        >
          <h1
            className="card-title title-l"
          >
            Semantic Resources
          </h1>
        </div>
        <div
          className="card-body"
        >
          <ul
            className="list list--highlighted"
          >
            <li
              className="tile tile--singleline"
            >
              <h4
                className="tile-maintext body-s"
              >
                /model/model.ecore
              </h4>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="card projectrepresentationslistcard"
      >
        <div
          className="card-header"
        >
          <h1
            className="card-title title-l"
          >
            Representations
          </h1>
        </div>
        <div
          className="card-body"
        >
          <ul
            className="list list--highlighted"
          >
            <li
              className="tile tile--twoline"
            >
              <div>
                <h4
                  className="tile-maintext body-s"
                >
                  Class Diagram
                </h4>
                <h5
                  className="tile-additionaltext caption-xs"
                />
              </div>
            </li>
            <li
              className="tile tile--twoline"
            >
              <div>
                <h4
                  className="tile-maintext body-s"
                >
                  Activity Diagram
                </h4>
                <h5
                  className="tile-additionaltext caption-xs"
                />
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      className="projectview-workflow"
    >
      <div
        className="card workflowcard"
      >
        <div
          className="card-header"
        >
          <h1
            className="card-title title-l"
          >
            Workflow
          </h1>
        </div>
        <div
          className="card-body"
        >
          <div
            className="tabbar"
          >
            <div
              className="tabbar-nav tabbar-nav--disabled"
              data-side="previous"
              onClick={[Function]}
            >
              &lt;
            </div>
            <div
              className="tabbar-tabs"
            />
            <div
              className="tabbar-nav tabbar-nav--disabled"
              data-side="next"
              onClick={[Function]}
            >
              &gt;
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
